### PR TITLE
#167729796 Implement Notification feature

### PR DESCRIPTION
**Pull request by okafor-chidimma**
#### What does this PR do?
Implementation of the Notification Feature

#### Description of Task to be completed?
Upon successful booking of a trip, a user should get an email notification and the user's manager should get an in-app notification of the trip

#### How should this be manually tested?
This feature is dependent on the create-trip route, so to test it follow these steps carefully

- Pull from the branch
- Navigate to **localhost:3000//api/v1/trips/get_trips/:id**  where **id** is the manager Id of the user that wants to create trip
- Navigate to the login route from your postman and Login
- Copy and keep the token
- Navigate to **http://localhost:3000/api/v1/trips/request** and pass in the trip details
- Ensure to pass in the **Bearer token** in the header of the request as **Authorization**
Pr

- Ensure to have correct info in the details
PR-prrof

- You should receive a success message
- Check your email to see the trip confirmation
- Go back to your browser to confirm that the trip details are showing for the manager without refreshing

#### Any background context you want to provide?
Ensure that the create-trip feature has been merged before testing

#### What are the relevant pivotal tracker stories?
#167729796

#### Screenshots (if appropriate)
proof_notify
proof_notify2

#### Questions:
None

### CODE REVIEW

File: src/controllers/signup.js [hunk 1]
```diff
@@ -62,6 +62,7 @@ export const signup = async (req, res) => {
       }
     });
   } catch (err) {
+    console.log(err, 'error from sign up');
```
Comment (sebaztyn) [inline src/controllers/signup.js]: Kindly take down console.log statements after debugging

File: src/routes/index.js [hunk 2]
```diff
@@ -3,6 +3,6 @@ import apiRouter from './api';
 
 const index = express.Router();
 
-index.use('/api', apiRouter);
+const foo = (io = null) => index.use('/api', apiRouter(io));
 
-export default index;
+export default foo;
```
Comment (sebaztyn) [inline src/routes/index.js]: Is there a reason why you used 'foo' a variable name? I think a name that helps understand what the we seek to achieve might be better

File: src/test/signup.test.js [hunk 3]
```diff
@@ -387,6 +388,7 @@ describe('POST signup route', () => {
           done();
         });
     } catch (err) {
+      console.log(err, 'error from test');
```
Comment (sebaztyn) [inline src/test/signup.test.js]: please take down the console.log after debugging

File: src/test/signup.test.js [hunk 4]
```diff
@@ -413,6 +415,7 @@ describe('POST signup route', () => {
...
```
Comment (sebaztyn) [inline src/test/signup.test.js]: please take down the console.log after debugging

**Review by sebaztyn — CHANGES_REQUESTED** [review]

**Review by chibytez — APPROVED** [review]
Nice work

**Review by mthamayor — APPROVED** [review]

**Review by sebaztyn — APPROVED** [review]
I tested locally and on postman. It works well

File: src/controllers/Trips.js [hunk 5]
```diff
@@ -1,8 +1,31 @@
...
+
+import autoBind from 'auto-bind';
+
+import notifyUserService from '../services/notifyUserService';
+
 import response from '../utils/Response';
 
-const { serverResponse } = response;
+import passwordEmail from '../utils/Mailer';
+
```
Comment (Stylll) [inline src/controllers/Trips.js]: can we remove the unnecessary spaces between the imports?

File: src/database/migrations/20190901120620-adding_to_user.js [hunk 6]
```diff
@@ -26,7 +26,10 @@ export default {
         'Developer',
         'QA',
         'Maintenance',
+        'User',
```
Comment (Stylll) [inline src/database/migrations/20190901120620-adding_to_user.js]: We should try as much as possible to not edit an existing migration. You should move these changes to a new migration file

File: src/database/migrations/20190901214649-create-notification.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      primaryKey: true
+    },
+    tripId: {
+      type: Sequelize.STRING,
+      references: {
+        model: 'Requests',
+        key: 'tripId'
+      }
+    },
+    lineManager: {
```
Comment (Stylll) [inline src/database/migrations/20190901214649-create-notification.js:17]: This should be a foreign key referencing the users table

File: src/database/migrations/20190901214649-create-notification.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      type: Sequelize.INTEGER,
+      allowNull: false
+    },
+    userId: {
+      type: Sequelize.STRING,
```
Comment (Stylll) [inline src/database/migrations/20190901214649-create-notification.js:22]: If you are saving the userId here, (as the id column on the users table), then this should be a foreign key referencing the users table

File: src/index.js [hunk 9]
```diff
@@ -19,6 +21,8 @@ const isProduction = process.env.NODE_ENV === 'production';
 
 // Create global app object
 const app = express();
+const server = http.Server(app);
+export const io = SocketIO(server);
```
Comment (Stylll) [inline src/index.js]: Is there a reason we are using `server` now instead of `app`?

File: src/middlewares/passwordResetMiddleware.js [hunk 10]
```diff
@@ -62,6 +62,17 @@ class authenticator {
...
+  /**
+ *
+ *
+ * @static
+ * @param {object} req
+ * @param {object} res
+ * @param {method} next
+ * @returns {Method} Calls next method on success
+ * @memberof authenticator
+ */
```
Comment (Stylll) [inline src/middlewares/passwordResetMiddleware.js]: Is that necessary here?

File: src/services/notifyUserService.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+ *
+ *
+ * @static
+ * @param {Integer} id
+ * @param {String} updatePassword
+ * @returns {Object} User
+ * @memberof userService
+ */
+  static async getSpecificTrips(id) {
```
Comment (Stylll) [inline src/services/notifyUserService.js]: I don't think this method is related to the NotificationService hence it shouldn't be in this class

File: src/routes/api/trips.route.js [hunk 12]
```diff
@@ -3,12 +3,18 @@ import { Trips } from '../../controllers';
...
 
-const { multiCityRequest } = Trips;
 const { authenticate } = Authentication;
 
 
 const router = express.Router();
 
-router.post('/request', authenticate, multicityCheck, validateInput, multiCityRequest);
+const tripRoute = (io = null) => {
+  const notify = new Trips(io);
```
Comment (Stylll) [inline src/routes/api/trips.route.js]: the variable name here doesn't correlate with the class. The class of `Trips` is now `notify`?

File: src/controllers/Trips.js [hunk 13]
```diff
@@ -19,9 +52,23 @@ class Trips {
  * @returns {object} Details of booked trips
  * @memberof Trips
  */
-  static async multiCityRequest(req, res, next) {
+  async multiCityRequest(req, res, next) {
```
Comment (Stylll) [inline src/controllers/Trips.js]: I noticed you changed the method to a non-static one and added a constructor to this controller. This controller should be for handling Trips related requests, not initializing `sockets.io`.

**Review by Stylll — CHANGES_REQUESTED** [review]
Hey @okafor-chidimma, I left a few comments. Kindly check them out.

File: src/database/migrations/20190901214649-create-notification.js (hunk 8 above)
Comment (okafor-chidimma) [reply src/database/migrations/20190901214649-create-notification.js:22]: resolved

**Review by sebaztyn — APPROVED** [review]

File: src/database/migrations/20190901214649-create-notification.js (hunk 7 above)
Comment (okafor-chidimma) [reply src/database/migrations/20190901214649-create-notification.js:17]: resolved

File: src/services/notifyUserService.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import Model from '../models';
+
+const { Notifications } = Model;
+/**
+ *
+ *
+ * @class userService
```
Comment (Stylll) [inline src/services/notifyUserService.js:7]: @okafor-chidimma jsdoc should be notifyUserService

**Review by Stylll — APPROVED** [review]

**Pull request merged**